### DISCUSSION TITLE
Retrying AMQPConnectionException

### DISCUSSION THREAD

**Discussion by jwage** [pdezwart/php-amqp#548] [Q&A]
If you get an AMQPConnectionException with the message "Library error: a SSL error occurred" or "a socket error occurred", is it safe to retry or in theory is it possible to get this exception after a message is successfully published on the server but before the response was received by the client?

I think if the timing is just right and you have retries implemented, then you could end up double publishing the message in failure scenarios.

Related to this change in Symfony https://github.com/symfony/symfony/pull/54167#issuecomment-2008929049

**Comment by tunterreitmeier** [comment]
This is just a limited perspective by having a look around the code for `publish` and `AMQPConnectionException` but without full knowledge of this extension and especially rabbitmq-c: the exception is only thrown for specific stati returned by the library client, which are all related to the connection to the server. So I think it's fair to assume that: "no connection -> nothing published"
Reference: https://github.com/php-amqp/php-amqp/blob/latest/amqp.c#L493

Bu this might also be of interest to you: we were running into the exact issue your contribution aims to fix (retry on lost connection). However, we are getting a `AMQPException` instead of `AMQPConnectionException`. Reason is the AWS Load Balancer which results in a "a SSL error occured". Same in this issue: https://github.com/symfony/symfony/issues/48241
That's why I created this PR here to also treat SSL and TCP errors as a `AMQPConnectionException`: https://github.com/php-amqp/php-amqp/pull/571
And thanks for the fix, by the way! These errors caused quite some headaches in our team.
